Issue 4140 add download button de

RomaDF opened the pull request.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/freelawproject/courtlistener?pullRequest=4334) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/freelawproject/courtlistener?pullRequest=4334) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/freelawproject/courtlistener?pullRequest=4334) it.</sub>

RomaDF commented:
Addressing issue: https://github.com/freelawproject/courtlistener/issues/4140
General comments/assumptions:
- I have added the class name to the columns in the csv. This is because we had repeated column names like id and description
- For the file name, the issues says: $slugifiedCaseName.$courtId.$courtlistenerID.$generation-date.csv. For example: national-veterans-legal-services-program-v-united-states.dcd.4214664.2024-06-26.csv
For the $courtlistenerID I understood docket id based on the example
- When creating the csv I have added the options quotechar'"' and quoting=csv,QUOTE_ALL. This means that every field in the csv will be converted to char and enclosed in ".
[screen_shot_.tar.gz](https://github.com/user-attachments/files/16696574/screen_shot_.tar.gz)

mlissner commented on cl/opinion_page/urls.py:
```diff
@@ -54,6 +55,11 @@
         "docket/<int:pk>/<blank-slug:slug>/", view_docket, name="view_docket"  # type: ignore[arg-type]
     ),
+    path(
+        "docket/download/<int:docket_id>/",
```
I think, for better or worse, we have a pattern of doing this instead:

```suggestion
        "docket/<int:docket_id>/download/",
```

mlissner commented on cl/search/models.py:
I think this might make more sense in `model_helpers.py`?

RomaDF replied:
I have moved the Mixin to cl/lib/model_helpers.py. 
Is this what you meant?

mlissner replied:
Yep, exactly, thanks.

mlissner commented on cl/opinion_page/views.py:
```diff
@@ -563,6 +583,28 @@ async def make_thumb_if_needed(
+    case_name = docket.slug
+
+    date_str = datetime.datetime.now().strftime("%Y-%m-%d")
+    filename = f"{case_name}.{court_id}.{docket_id}_{date_str}.csv"
```
I think we wanted these to be dot-separated, so:

```suggestion
    filename = f"{case_name}.{court_id}.{docket_id}.{date_str}.csv"
```

RomaDF replied:
Yes! fixed

mlissner commented on cl/opinion_page/views.py:
```diff
@@ -563,6 +583,28 @@ async def make_thumb_if_needed(
+    date_str = datetime.datetime.now().strftime("%Y-%m-%d")
+    filename = f"{case_name}.{court_id}.{docket_id}_{date_str}.csv"
+
+    # TODO check if for large files we'll cache or send file by email
+    csv_content = await sync_to_async(generate_docket_entries_csv_data)(
+        de_list
+    )
```
Our biggest issue here will be putting the file into memory before serving it. I wonder if we could stream it to the user as we generate it. I *think* all the pieces are there for this *but* if we do this *and* want to cache to disk, we'll need to stream to S3 and to the user. Now we're into the fun stuff...

Maybe later. :)

mlissner commented on cl/opinion_page/views.py:
```diff
@@ -336,19 +340,34 @@ async def redirect_docket_recap(
+async def fetch_docket_entries(docket):
+    """Fetch docket entries asociated to docket
+
+    param request: current HttpRequest.
```
This looks out of date.

RomaDF replied:
Doc updated

mlissner commented on cl/opinion_page/views.py:
```diff
@@ -1,7 +1,10 @@
 from http import HTTPStatus
-from typing import Any, Dict, Union
+from io import StringIO
+from typing import Any, Dict, List, Union
```
Did you wind up using `List` or `StringIO` in this file?

RomaDF replied:
StringIO but import should not be in this file any longer, I have deleted them.

mlissner commented on cl/search/models.py:
```diff
@@ -1135,8 +1135,51 @@ class Meta:
         proxy = True
 
 
+# @pghistory.track(AfterUpdateOrDeleteSnapshot(), obj_field=None)
```
We'll want to remove this comment, right?

RomaDF replied:
No, this is a Mixin not a model, if  I understood correctly the doc this decorator should be use in models. I have deleted the comment and I have moved this to model_helpers

mlissner reviewed (changes requested):
Thanks Roma! This is great. I have only done a quick review, but here are my thoughts:

1. In #4140, I said the button location should be...

    > Maybe just below the "Next" button at the top of the table, floated right?

    That wasn't very clear. What I meant was that it should go in the actual table of entries, more like this:

    ![image](https://github.com/user-attachments/assets/f50429ff-e5fb-42b3-b406-805043168f98)

2. There's a comment above about signing our CLA. Can you do that please, so we can merge? 

3. We'll want ratelimiting on this before we send it live. Maybe 1/m? We have a decorator for this.

4. I made a comment about caching vs. streaming. Fun stuff, but it can certainly wait until we try this and see what happens. Maybe it'll be fine.

5. I haven't looked at your HTML or tests, I'll let @ERosendo dig into those. 

----

But big picture, this looks great, thank you so much for putting in this effort.

But despite my review (and probably more coming from Eduardo), please don't feel ANY need to continue working on this. It can easily wait, and you've done the task at hand, which was to show us your skills. 

I think it's great and I'll be in touch by email with next steps! 

Thank you!

mlissner commented:
Oh, one last thing. An example of the CSV will be helpful before we merge to eyeball it and make sure it looks about right — but this can wait!

ERosendo commented:
I've been reviewing the code changes in this PR and here are some thoughts:

> What I meant was that it should go in the actual table of entries.

I believe this proposal makes sense. We can relocate the code for the button from the `de_filter.html` template to the `de_list.html` template. Here are some approaches I've explored for integrating the button into our responsive design:

![Screen Recording 2024-08-23 at 4 51 20 PM](https://github.com/user-attachments/assets/78e9aa03-81bb-4164-a7aa-ebaf502bb157)

![Screen Recording 2024-08-23 at 4 46 01 PM](https://github.com/user-attachments/assets/4cbae533-9699-4d46-8e31-64edd51c880a)

Adding 'csv' near the download icon offers a clear visual cue about the file format, but I'm uncertain about its responsiveness, @mlissner what do you think?

> An example of the CSV will be helpful before we merge to eyeball it and make sure it looks about right

@mlissner Here's an example of the CSV file:

[clarence-t-foos.ohnb.4826703.2024-08-23.csv](https://github.com/user-attachments/files/16733823/clarence-t-foos.ohnb.4826703.2024-08-23.csv)

I reviewed the file and noticed that the link in the`filepath_local` field is not a full URL to the storage location.

mlissner commented on cl/search/models.py:
```diff
@@ -1736,6 +1760,47 @@ def as_search_dict(self, docket_metadata=None):
+    def _get_readable_document_type(self, *args, **kwargs):
+        return self.get_document_type_display()
+
+    def _get_readable_ocr_status(self, *args, **kwargs):
+        return self.get_ocr_status_display()
```
Per Eduardo's comment, let's do one more of these for filepath_local. Notionally:

```
    def _get_full_filepath_local(self, *args, **kwargs):
        if self.filepath_local:
            return f"https://storage.courtlistener.com/{self.filepath_local}"
        return ""
```

RomaDF replied:
done

RomaDF replied:
[Uploading united-states-v-cadden.mad.4275782.2024-08-25.csv…]()
Explample with change in filepath if any

mlissner commented:
I like the second gif you posted, Eduardo, with the button. Looks great to me. 

The CSV looks pretty good too. Lots of fields in there that'll confuse people, but the important ones are there too, and we like to provide lots of info when we can, so I think it's good.

RomaDF commented:
I have changed the button to de_list. It's not as good as Eduardo's gift, though

ERosendo commented on cl/lib/model_helpers.py:
```diff
@@ -562,3 +562,46 @@ def linkify_orig_docket_number(agency: str, og_docket_number: str) -> str:
+            "Subclass must implement get_column_fuction method"
+        )
+
+    def to_csv_row(self) -> List[str]:
```
From Python 3.9 ([PEP 585](https://www.python.org/dev/peps/pep-0585/)) onwards `dict`, `list` and various other classes are now generic types. Using these rather than their typing counterpart is now preferred. From Python 3.9 you can now just do:

```suggestion
    def to_csv_row(self) -> list[str]:
```

ERosendo commented on cl/lib/model_helpers.py:
```diff
@@ -1,7 +1,7 @@
 import os
 import re
-from typing import Optional
+from typing import Callable, Dict, List, Optional
```
Importing `Dict` and `List` from typing is redundant. We can use built-in types (`dict` and `list`):

```suggestion
from typing import Callable, Optional
```

ERosendo commented on cl/opinion_page/views.py:
```diff
@@ -361,11 +377,12 @@ async def view_docket(
         if cd.get("filed_before"):
             de_list = de_list.filter(date_filed__lte=cd["filed_before"])
         if cd.get("order_by") == DocketEntryFilterForm.DESCENDING:
-            sort_order_asc = False
```
Removing the `sort_order_asc` variable from this line without providing a suitable alternative will negatively impact the functionality of the filter bar in the `de_filter.html` template. We use this variable to set the active state of the radio buttons, ensuring that the filter bar accurately reflects the current sorting order. 

<img width="1164" alt="Screenshot 2024-08-26 at 5 15 28 PM" src="https://github.com/user-attachments/assets/1c69c582-f3e9-4979-a16b-9e8eb8837234">

If removed, the 'asc' button will remain active even when the entries are sorted in descending order, leading to a confusing user experience.

RomaDF replied:
Sorry about this, it was not intentioanl. Fixed

ERosendo commented on cl/opinion_page/tests.py:
```diff
@@ -1523,3 +1533,145 @@ async def test_block_cluster_and_docket_via_ajax_view(self) -> None:
+        RECAPDocument.objects.all().delete()
+        User.objects.all().delete()
+
+    def test_fetch_docket_entries(self) -> None:
+        """Verify that fetch entries function returns right docket_entries"""
+        res = asyncio.run(fetch_docket_entries(self.mocked_docket))
+        self.assertEqual(len(res), len(self.mocked_docket_entries))
+        self.assertIn(self.mocked_docket_entries[0], res)
+        self.assertNotIn(self.mocked_extra_docket_entries[0], res)
```
We can use asynchronous alternatives and refactor this test to avoid importing `asyncio`. 

```suggestion
    async def test_fetch_docket_entries(self) -> None:
        """Verify that fetch entries function returns right docket_entries"""
        res = await fetch_docket_entries(self.mocked_docket)
        self.assertEqual(await res.acount(), len(self.mocked_docket_entries))
        self.assertTrue(await res.acontains(self.mocked_docket_entries[0]))
        self.assertFalse(
            await res.acontains(self.mocked_extra_docket_entries[0])
        )
```

ERosendo commented on cl/opinion_page/tests.py:
```diff
@@ -1523,3 +1533,145 @@ async def test_block_cluster_and_docket_via_ajax_view(self) -> None:
+
+    def test_generate_docket_entries_csv_data(self) -> None:
+        """Verify str with csv data is created. Check column and data entry"""
+        res = generate_docket_entries_csv_data(self.mocked_docket_entries)
+        res_lines = res.split("\r\n")
+        res_line_data = res_lines[1].split(",")
+        self.assertEqual(res[:16], '"docketentry_id"')
+        self.assertEqual(res_line_data[1], '"506581111"')
+
+    def test_view_download_docket_entries_csv(self) -> None:
+        """Test download_docket_entries_csv returns csv content"""
+        with mock.patch(
+            "cl.opinion_page.utils.generate_docket_entries_csv_data"
+        ) as mock_download_function:
+            with mock.patch(
+                "cl.opinion_page.utils.core_docket_data"
+            ) as mock_core_docket_data:
+                with mock.patch(
+                    "cl.opinion_page.utils.user_has_alert"
+                ) as mock_user_has_alert:
+                    mock_download_function.return_value = (
+                        '"col1","col2","col3"\r\n"value1","value2","value3"'
+                    )
+                    mock_user_has_alert.return_value = False
+                    mock_core_docket_data.return_value = (
+                        self.mocked_docket,
+                        {
+                            "docket": self.mocked_docket,
+                            "title": "title",
+                            "note_form": "note_form",
+                            "has_alert": mock_user_has_alert.return_value,
+                            "timezone": "EST",
+                            "private": True,
+                        },
+                    )
+
+                    response = async_to_sync(download_docket_entries_csv)(
+                        self.request, self.mocked_docket.id
+                    )
+                    self.assertEqual(response["Content-Type"], "text/csv")
```
While mocking is a great testing strategy, nested context managers can sometimes hinder code clarity. In this instance, using the `@mock.patch` decorator provides a more concise and readable approach to achieving the same goals.

```python
    @mock.patch("cl.opinion_page.utils.user_has_alert")
    @mock.patch("cl.opinion_page.utils.core_docket_data")
    @mock.patch("cl.opinion_page.utils.generate_docket_entries_csv_data")
    def test_view_download_docket_entries_csv(
        self,
        mock_download_function,
        mock_core_docket_data,
        mock_user_has_alert,
    ) -> None:
        """Test download_docket_entries_csv returns csv content"""
        mock_download_function.return_value = (
            '"col1","col2","col3"\r\n"value1","value2","value3"'
        )
        mock_user_has_alert.return_value = False
        mock_core_docket_data.return_value = (
            self.mocked_docket,
            {
                "docket": self.mocked_docket,
                "title": "title",
                "note_form": "note_form",
                "has_alert": mock_user_has_alert.return_value,
                "timezone": "EST",
                "private": True,
            },
        )

        response = async_to_sync(download_docket_entries_csv)(
            self.request, self.mocked_docket.id
        )
        self.assertEqual(response["Content-Type"], "text/csv")
```

RomaDF replied:
It makes sense. Done as suggested

ERosendo commented:
@RomaDF Thanks for your contributions to the pull request! I've added code to enhance the responsive design of the "export CSV" button. 👍 

@mlissner  I've added comments to enhance the code of the new tests, and refine type annotations and docstrings.

Before merging the code, it's important to address the point raised in this comment: 

https://github.com/freelawproject/courtlistener/pull/4334#discussion_r1731839531

RomaDF commented on cl/opinion_page/views.py:
```diff
@@ -1,4 +1,6 @@
+import csv
 import datetime
+import logging
```
done

RomaDF commented on cl/opinion_page/views.py:
```diff
@@ -336,19 +339,32 @@ async def redirect_docket_recap(
+    """Fetch docket entries asociated to docket
+
+    param docket: docket.id to get related docket_entries.
+    returns: DocketEntry list.
```
done
